streams/src/test/java/org/apache/kafka/streams/integration/StandbyTaskEOSIntegrationTest.java

```diff
@@ -170,8 +170,8 @@ public void shouldSurviveWithOneTaskAsStandby() throws Exception {
         // Wait for the record to be processed
         assertTrue(instanceLatch.await(15, TimeUnit.SECONDS));
 
-        streamInstanceOne.close(Duration.ZERO);
-        streamInstanceTwo.close(Duration.ZERO);
```

Comment:
This test failing does give me some small pause on just throwing an IllegalStateException in case of PENDING_SHUTDOWN/ERROR -- do we think many users follow this pattern of calling the non-blocking `close` and then immediately calling `cleanUp()`? Maybe the better question is, should we support this, or is it alright to just enforce that users wait for Streams to shut down before invoking `cleanUp` -- if we do want to support this, we could just signal to the shutdown thread that it should re-invoke `cleanUp` once Streams finished shutting down, instead of just throwing an IllegalStateException. Not sure if it's really worth it or not, though -- thoughts @guozhangwang  @wcarlson5 ?

Reply:
How hard could it be to have `cleanUp()` check set a flag  for the shutdown thread to call clean up after it's done? We want to make sure the shutdown thread checks after it changes the state to `NOT_RUNNING` or `ERROR`.

I think this a much more flexible pattern that would be good to support if it's running or rebalancing we would still want to throw and `IllegalStateException` or maybe `IllegalOperationException`. But either way it would be good to test when stream is in a` PENDING` state. But I think you would have to make that a unit test

Reply:
I already have a unit test in this PR for the `PENDING` case, I should be able to adapt that to the new behavior without much trouble, if we want to change it. I also don't think it would be that difficult to implement this behavior -- I guess it was more a question of whether we should handle this semi-transparently or just tell the user to be patient and try again after the close. For example, at the moment `close()` will launch the shutdown helper as a daemon thread, so if the user does a non-blocking close followed by `cleanUp` and that's the end of the app, most likely the clean up will not occur (nor will the app complete its shutdown)

Reply:
I guess another alternative is just to have `cleanUp` block until Streams finishes shutting down, but I don't think we want to make it essentially block on the results of `close()`, at least not without giving users a way to supply a timeout, and I don't think it's worth doing an entire KIP just to add an overload with one

Reply:
I think if they call a non blocking close then clean up then later notice the leftovers that is not a great experience and if can prevent it we should. I agree probably not worth a KIP though

Reply:
At the moment, if user calls a close() with no blocking and then cleanUp, the latter `cleanUp` could become a no-op because the locks are not released yet, and user won't know; whereas now it would throw an exception saying that close is not done yet, right? Plus our javadoc of `cleanUp` states that

```
     * Do a clean up of the local {@link StateStore} directory ({@link StreamsConfig#STATE_DIR_CONFIG}) by deleting all
     * data with regard to the {@link StreamsConfig#APPLICATION_ID_CONFIG application ID}.
     * <p>
     * May only be called either before this {@code KafkaStreams} instance is {@link #start() started} or after the
     * instance is {@link #close() closed}.
```
Hence I think the new behavior is not really a regression but really a fix forward.